docs/design/marketing-offer-program.md

```diff
@@ -67,54 +67,164 @@ The credit is automatically applied to the account and will have a max limit tha
     created_at - timestamp
     referred_by - bytea (nullable)
     FOREIGN KEY (offer_id)
     FOREIGN KEY (referred_by)
-    PRIMARY KEY (user_id)
+    FOREIGN KEY (user_id)
+    PRIMARY KEY (id)
 ```
 
 **user table**
 ```sql
     total_Referred - int
 ```
 
-### Offer Program Service
+### Marketing Service
 
-**satellite/marketing/offers.go**
-- Create offers interface to interact with offer table
+**satellite/marketing/service.go**
+```golang
+func (m *marketing) GetCurrentOffer(ctx context.Context, offerStatus OfferStatus) (*Offer, error) {
+  isDefault := offerStatus == Default
+  offer, err := m.db.Marketing().Offers().GetCurrentOffer(ctx, isDefault)
+  if err != nil {
+    return nil, Error.Wrap(err)
+  }
+
+  return offer, nil
+}
+
+func (m *marketing) StopOffer(ctx context.Context, offerId Offer.ID) error {
+  o := UpdateOffer{
+    Status: Done,
+    ExpiresAt: time.Now(),
+  }
+  err := m.db.Marketing().Offers().UpdateOfferByID(ctx, offerId, o)
+  if err != nil {
+    return Error.Wrap(err)
+  }
+
+  return nil
+}
+
+func (m *marketing) Create(ctx context.Context, offer Offer) error {
+  if offer.Status == Default {
+    offer.ExpiresAt = time.Now().AddDate(100, 0, 0)
+  }
+
+  err := m.db.Marketing().Offers().Create(ctx, offer)
+  if err != nil {
+    return Error.Wrap(err)
+  }
+
+  return nil
+}
 
+func (m *marketing) ListAllOffers(ctx context.Context) ([]Offers, error)
+```
+
+**satellite/marketing/offers.go**
 ```golang
-type DB interface {
+type Offers interface {
   ListAllOffers(ctx context.Context) ([]Offer, error)
-  GetOfferById(ctx context.Context, offerId Offer.ID) (Offer, error)
-  Update(ctx context.Context, offer *Offer) (*Offer, error)
-  Delete(ctx context.Context, offerId Offer.ID)
+  GetCurrentOffer(ctx context.Context, isDefault bool) (Offer, error)
   Create(ctx context.Context, offer *Offer)
 }
 
-func (offer *Offer) IsUpdatable() bool {
-  return offer.Status == NO_STATUS
+type OfferStatus int
+const (
+  Done OfferStatus = 0
+  Active OfferStatus = 1
+  Default OfferStatus = 2
+)
+
+type UpdateOffer struct {
+  Status OfferStatus
+  ExpiresAt time.time
+  NumRedeemed int
+}
+```
+
+**satellite/satellitedb/offers.go**
+```golang
+func (o *offersDB) GetCurrentOfferByStatus(ctx context.Context, isDefault bool) (*marketing.Offer, error) {
+  var statement string
+  if isDefault {
+    statement = `
+      SELECT * FROM offers WHERE offers.status=? AND offers.expires_at>NOW() order by offers.created_at desc;
+    `
+  } else {
+    statement := `WITH o AS (
+        SELECT * FROM offers WHERE offers.status=? AND offers.expires_at>NOW()
+      )
+      SELECT * FROM o
+      UNION ALL
+      SELECT * FROM offers
+      WHERE offers.status=?
+      AND NOT EXISTS (
+      SELECT * FROM o
+      ) order by offers.created_at desc;`
+  }
+
+  currentOffer := returnedValueWithLatestCreatedAt
+  return currentOffer, nil
+}
+func (o *offersDB) Create(ctx context.Context, offer *marketing.Offer) error {
```

Comment:
On a higher level, could you explain what this function is doing? It is a little difficult for me to decipher what it's doing and why just by looking at the code.

Reply:
Sorry for the confusion
```golang
// First check if there is an `active` or `default` offer
// If there's one, we need to update its status to be `Done`
// then we will create the new offer
```

Reply:
Ok that makes sense